tika-parsers/src/main/java/org/apache/tika/parser/html/charsetdetector/PreScanner.java

```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * Read an attribute from the stream
+     *
+     * @return the attribute as a Map.Entry, where the key is the attribute's name and
+     * the value is the attribute's value. If there is no attribute, return null
+     */
+    private Map.Entry<String, String> getAttribute() throws IOException {
+        String name = getAttributeName();
+        if (name == null) return null;
+
+        if (!expect(EQUAL)) return new AbstractMap.SimpleEntry<>(name, "");
+        skipAll(WHITESPACE);
+
+        String value = getAttributeValue();
+        return new AbstractMap.SimpleEntry<>(name, value);
+    }
+
+    private String getAttributeName() throws IOException {
+        skipAll(SPACE_OR_SLASH);
+        if (expect(TAG_END)) return null;
+        StringBuilder name = new StringBuilder();
+        while (!(peek() == EQUAL && name.length() > 0) &&
+                !(peek() == TAG_END || peek() == SLASH) &&
+                !skipAll(WHITESPACE)) {
+            name.append((char) getLowerCaseChar());
+        }
+        return name.toString();
+    }
+
+    private String getAttributeValue() throws IOException {
+        StringBuilder value = new StringBuilder();
+        stream.mark(1);
+        byte quote = read();
+        if (contains(QUOTE, quote)) {
+            for (byte b = getLowerCaseChar(); b != quote; b = getLowerCaseChar()) {
+                value.append((char) b);
+            }
+        } else {
+            stream.reset();
+            for (byte b = getLowerCaseChar(); !contains(SPACE_OR_TAG_END, b); b = getLowerCaseChar()) {
+                value.append((char) b);
+                stream.mark(1);
+            }
+            stream.reset(); // unread the space or tag end
+        }
+        return value.toString();
+    }
+
+    private boolean skipAll(BitSet bitSet) throws IOException {
+        boolean skipped = false;
+        stream.mark(1);
+        for (byte read = read(); contains(bitSet, read); read = read()) {
+            skipped = true;
+            stream.mark(1);
+        }
+        stream.reset();
+        return skipped;
+    }
+
+    private byte getLowerCaseChar() throws IOException {
+        byte nextPoint = read();
+        if (nextPoint >= 'A' && nextPoint <= 'Z') nextPoint += 0x20; // lowercase
+        return nextPoint;
+    }
+
+    private boolean processComment() throws IOException {
+        if (!expect(COMMENT_START)) return false;
+        if (!expect(TAG_END)) skipUntil(COMMENT_END);
+        return true;
+    }
+
+    private boolean expect(byte... expected) throws IOException {
+        stream.mark(expected.length);
+        for (byte b : expected) {
+            byte read = read();
+            if (read != b) {
+                stream.reset();
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void skipUntil(byte... expected) throws IOException {
+        while (!expect(expected)) {
+            if (stream.skip(1) == 0) return;
```

Comment:
Depending on the stream type, you can't actually rely on the return value of `skip`.  I've learned the hard way to follow commons-io's `skipFully` and use `read` instead.